### DISCUSSION TITLE
가비지 컬렉션에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#84] [Java]
설명해주세요 제발!

**Comment by kunsanglee** [comment]
가비지 컬렉션(Garbage Collection, GC)은 더 이상 사용되지 않는 메모리를 자동으로 찾아 해제하는 프로세스입니다. 개발자가 직접 메모리를 관리해야 하는 부담을 줄여주고, 메모리 누수(memory leak)와 같은 문제를 방지하는데 도움을 줍니다.

### 가비지 컬렉션 작동 원리
가비지 컬렉션의 기본 원리는 루트 집합(root set)으로부터 시작하여 접근 가능한 객체들을 추적합니다. 루트 집합은 메서드 영역의 클래스, 스택 영역의 지역 변수와 매개변수, 네이티브 메서드 스택 등에서 참조되는 객체들을 포함합니다. 이러한 루트 집합에서부터 시작하여 참조를 따라가며, 접근 가능한(사용중인) 객체를 식별합니다. 어떤 루트 집합에서도 접근 불가능한 객체(사용중이지 않은)들을 가비지로 간주되어 메모리 해제 대상이 됩니다.

### 가비지 컬렉션 종류
Minor GC는 새로 생성된 대부분의 객체가 위치하는 Young 영역에서 더 이상 사용되지 않는 객체를 정리합니다. Young 영역은 Eden, Survivor 영역으로 나뉩니다.
Major GC(Full GC)는 Young 영역 뿐만 아니라 객체들이 장기간 저장되는 Old 영역, Method 영역에서 사용되지 않는 객체를 정리합니다. Major GC는 일반적으로 더 시간이 많이 걸리며 애플리케이션의 작동을 멈출 수도 있습니다.

**Comment by seunghye218** [comment]
가비지 컬렉션은 프로그램에서 더 이상 사용하지 않는 객체가 차지하는 메모리를 회수하는 자동화된 메모리 관리 시스템입니다. 복잡하고 오류가 발생하기 쉬운 메모리 할당 및 해제 부담을 덜어줍니다.

가비지 컬렉션은 특정 조건이 충족되면 자동으로 수행됩니다.  
첫 번째로 힙에서 사용 가능한 여유 메모리가 특정 임계값 아래로 떨어지는 경우,  
두 번째로 새 객체가 생성되고 이를 수용할 공간이 힙에 충분하지 않은 경우,  
마지막으로 일반적으로 권장되지는 않지만 `System.gc()` 메서드를 사용하는 것 입니다.

가비지 컬렉션이 실행되면 프로그램 실행을 일시적으로 중지합니다. 너무 자주 또는 길면 성능 문제가 발생할 수 있습니다.
